massa-network-worker/src/network_worker.rs

```diff
@@ -96,55 +109,119 @@ impl NetworkWorker {
+        let (handshake_tx, handshake_rx) =
+            bounded::<(ConnectionId, HandshakeWorker)>(cfg.handshake_manager_channel_size);
+        let handshake_manager_join_handle =
+            start_handshake_manager(handshake_rx, conn_tx, runtime.handle().clone());
+
+        let (node_result_tx, node_result_rx) = bounded::<(
+            NodeId,
+            Result<ConnectionClosureReason, NetworkError>,
+        )>(cfg.node_result_channel_size);
+
+        let (handshake_peer_list_tx, handshake_peer_list_rx) =
+            bounded::<IpAddr>(cfg.handshake_peer_list_channel_size);
+
+        let (node_event_tx, node_event_rx) = bounded::<NodeEvent>(cfg.node_event_channel_size);
         let max_wait_event = cfg.max_send_wait_network_event.to_duration();
         NetworkWorker {
             cfg,
             self_node_id,
             keypair,
-            listener,
+            listener: Some(listener),
             establisher,
             peer_info_db,
             controller_command_rx,
             event: EventSender::new(controller_event_tx, node_event_tx, max_wait_event),
             controller_manager_rx,
             running_handshakes: HashSet::new(),
-            handshake_futures: FuturesUnordered::new(),
-            handshake_peer_list_futures: FuturesUnordered::new(),
+            handshake_peer_list_futures: Default::default(),
             node_event_rx,
             active_nodes: HashMap::new(),
-            node_worker_handles: FuturesUnordered::new(),
             active_connections: HashMap::new(),
             version,
+            runtime,
+            connections_rx,
+            handshake_tx,
+            handshake_manager_join_handle,
+            node_result_rx,
+            node_result_tx,
+            handshake_peer_list_tx,
+            handshake_peer_list_rx,
         }
     }
 
-    /// Runs the main loop of the network worker
-    /// There is a `tokio::select!` inside the loop
-    pub async fn run_loop(mut self) -> Result<(), NetworkError> {
-        let mut out_connecting_futures = FuturesUnordered::new();
+    /// Runs the main loop of the network worker.
+    pub fn run_loop(mut self) -> Result<(), NetworkError> {
+        let mut out_connection_handle = None;
+        let mut out_connection_rx = None;
         let mut cur_connection_id = ConnectionId::default();
 
-        // wake up the controller at a regular interval to retry connections
-        let mut wakeup_interval = tokio::time::interval(self.cfg.wakeup_interval.to_duration());
-        let mut need_connect_retry = true;
+        // Start a task to run the listener.
+        // TODO: config size.
+        let (listener_tx, listener_rx) = bounded::<(ReadHalf, WriteHalf, SocketAddr)>(1);
+        let mut listener = self
+            .listener
+            .take()
+            .expect("No listener at start of run loop.");
+        let listener_handle = self.runtime.spawn(async move {
```

Comment:
Maybe add a comment to tell that in more long term we to have it in a thread follow the discussion https://github.com/massalabs/massa/discussions/3149.

Because here you are on th same runtime as the rest of the network worker right ?

Reply:
I don't think we need to add a comment to that effect now, since it's not clear exactly how this will evolve.